File: spadea-android/app/src/main/java/com/spadea/xqipao/ui/base/view/BaseDialogFragment.java
package com.spadea.xqipao.ui.base.view;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.spadea.xqipao.ui.base.presenter.IPresenter;

import butterknife.ButterKnife;
import butterknife.Unbinder;


public abstract class BaseDialogFragment<P extends IPresenter> extends DialogFragment implements IView<FragmentActivity> {
    private static final String TAG = "BaseDialogFragment";

    protected P MvpPre;
    protected View rootView;
    protected Context mContext;
    protected Unbinder mUnbinder;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MvpPre = bindPresenter();
    }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
//        rootView = super.onCreateView(inflater, container, savedInstanceState);
        if (rootView == null) {
            rootView = LayoutInflater.from(getActivity()).inflate(getLayoutId(), container, false);
            mContext = getContext();
            mUnbinder = ButterKnife.bind(this, rootView);
            initView(rootView);
            initData();
            initListener();
        } else {
            mUnbinder = ButterKnife.bind(this, rootView);
            ViewGroup parent = (ViewGroup) rootView.getParent();
            if (parent != null) {
                parent.removeView(rootView);
            }
        }
        return rootView;
    }


    protected void initListener() {

    }


    protected abstract void initData();

    protected abstract void initView(View rootView);

    protected abstract int getLayoutId();

    protected abstract P bindPresenter();

    @Override
    public FragmentActivity getSelfActivity() {
        return getActivity();
    }

    @Override
    public void showLoadings() {

    }

    @Override
    public void disLoadings() {

    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        /**
         * 在生命周期结束时，将presenter与view之间的联系断开，防止出现内存泄露
         */
        mUnbinder.unbind();
        if (MvpPre != null) {
            MvpPre.detachView();
        }
    }


    public void show(FragmentManager fragmentManager) {
        show(fragmentManager, TAG);
    }

}

File: spadea-android/turntable/src/main/java/com/yutang/game/turntable/base/BaseDialogFragment.kt
package com.yutang.game.turntable.base

import android.app.Application
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseDialogFragment : DialogFragment() {

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        setCustomDensity(this)
        removeBg()
        return LayoutInflater.from(context).inflate(getLayoutId(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        dialog.setCanceledOnTouchOutside(true)
        initPresenter()
        initView()
        initListener()
        initData()
    }

    abstract fun initPresenter()

    abstract fun getLayoutId(): Int

    abstract fun initView()

    open fun initListener() {

    }

    open fun initData() {

    }

    private fun removeBg() {
        dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

    }

    override fun onStart() {
        super.onStart()

    }

    override fun onStop() {
        super.onStop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
        removeCustom()
    }

    open fun show(fragmentManager: FragmentManager?) {
        fragmentManager?.let {
            show(it, javaClass.name)
        }
    }

    /**
     * 今日头条适配方案
     * 以宽度375dp为设计图尺寸
     */
    var sNoncompatDensity: Float = 0f
    var sNoncompatScaledDensity: Float = 0f
    var oldDensityDpi: Int = 0

    fun setCustomDensity(fragment: Fragment) {
        val appDisplayMetrics = context?.applicationContext?.resources?.displayMetrics
        appDisplayMetrics?.let {
            if (sNoncompatDensity == 0f) {
                sNoncompatDensity = it.density
                sNoncompatScaledDensity = it.scaledDensity
                oldDensityDpi = it.densityDpi

                val targetDensity = it.widthPixels / 375f //关键代码
                val targetScaledDensity = targetDensity * (sNoncompatScaledDensity / sNoncompatDensity)
                val targetDensityDpi = 160 * targetDensity

                val fragmentDisplayMetrics = fragment.resources.displayMetrics
                fragmentDisplayMetrics.density = targetDensity
                fragmentDisplayMetrics.scaledDensity = targetScaledDensity
                fragmentDisplayMetrics.densityDpi = targetDensityDpi.toInt()
            }
        }
    }

    fun removeCustom() {
        val fragmentDisplayMetrics = resources.displayMetrics
        fragmentDisplayMetrics.density = sNoncompatDensity
        fragmentDisplayMetrics.scaledDensity = sNoncompatScaledDensity
        fragmentDisplayMetrics.densityDpi = oldDensityDpi
    }

}